samples/msal-node-samples/RedisTestApp/src/AuthProvider.ts

```diff
@@ -0,0 +1,411 @@
+        req.session.nonce = this.cryptoProvider.createNewGuid();
+
+        /**
+         * The OAuth 2.0 state parameter can be used to encode information of the app's state before redirect. 
+         * You can pass the user's state in the app, such as the page or view they were on, as input to this parameter. 
+         * MSAL allows you to pass your custom state as state parameter in the request object. For more information, visit:
+         * https://docs.microsoft.com/azure/active-directory/develop/msal-js-pass-custom-state-authentication-request
+         */
+        const state = this.cryptoProvider.base64Encode(
+            JSON.stringify({
+                stage: constants.AppStages.SIGN_IN,
+                path: req.route.path,
+                nonce: req.session.nonce
+            })
+        );
+
+        const params: AuthCodeParams = {
+            authority: this.getMsalConfig(req).auth.authority,
+            scopes: OIDC_DEFAULT_SCOPES,
+            state: state,
+            redirect: this.appSettings.settings.redirectUri,
+            prompt: PromptValue.SELECT_ACCOUNT,
+        };
+
+        // initiate the first leg of auth code grant to get token
+        this.getAuthCode(req, res, next, params);
+    };
+
+    /**
+     * Initiate sign out and clean the session
+     * @param {Request} req: express request object
+     * @param {Response} res: express response object
+     * @param {NextFunction} next: express next 
+     */
+    signOut = async (req: Request, res: Response, next: NextFunction): Promise<any> => {
+
+        /**
+         * Construct a logout URI and redirect the user to end the 
+         * session with Azure AD/B2C. For more information, visit: 
+         * (AAD) https://docs.microsoft.com/azure/active-directory/develop/v2-protocols-oidc#send-a-sign-out-request
+         * (B2C) https://docs.microsoft.com/azure/active-directory-b2c/openid-connect#send-a-sign-out-request
+         */
+        const logoutURI = `${this.getMsalConfig(req).auth.authority}/oauth2/v2.0/logout?post_logout_redirect_uri=${this.appSettings.settings.postLogoutRedirectUri}`;
+
+        req.session.isAuthenticated = false;
+
+        req.session.destroy(() => {
+            res.redirect(logoutURI);
+        });
+    }
+
+    /**
+     * Middleware that handles redirect depending on request state
+     * There are basically 2 stages: sign-in and acquire token
+     * @param {Request} req: express request object
+     * @param {Response} res: express response object
+     * @param {NextFunction} next: express next 
+     */
+    handleRedirect = async (req: Request, res: Response, next: NextFunction): Promise<any> => {
+
+        if (req.query.state) {
+            const state = JSON.parse(this.cryptoProvider.base64Decode(req.query.state as string));
+
+            // check if nonce matches
+            if (state.nonce === req.session.nonce) {
+                const msalClient = this.initializeMsalClient(req);
```

Comment:
So the trick is you're just creating a new ConfidentialClientApplication per request, instead of just once at startup. I agree that this solves the race condition though I'm a little concerned about the overhead of this. It doesn't seem like the way this api was designed to be used... certainly the in memory cache in here will be wiped and you'll have to fetch from redis every time.

But basically this would work, it just seems awfully hacky compared to just implementing the IStoragePlugin which really seems like the right way.

Reply:
First thank you for your review on this @justinmchase, really do appreciate the extra set of eyes on this feature. 

We are reviewing this with the team, to discuss whether there's a better way to implement this and not break our current APIs